qa/tasks/quiescer.py

```diff
@@ -0,0 +1,269 @@
+            quiesce_factor = self.MIN_QUIESCE_FACTOR
+
+
+        self.quiesce_factor = max(0.01, quiesce_factor)
+        self.min_quiesce = min_quiesce
+        self.max_quiesce = max_quiesce
+        self.initial_delay = initial_delay
+
+    def next_quiesce_duration(self):
+        mu = (self.min_quiesce + self.max_quiesce) / 2
+        sigma = 3 * math.sqrt(self.max_quiesce - self.min_quiesce)
+        duration = round(self.rnd.gauss(mu, sigma), 1)
+        duration = max(duration, self.min_quiesce)
+        duration = min(duration, self.max_quiesce)
+        return duration
+
+    def tell_quiesce_leader(self, *args):
+        leader = None
+        rc = None
+        stdout = None
+
+        while leader is None and not self.is_stopped:
+            leader = self.fs.get_var('qdb_leader')
+            if leader is None:
+                self.logger.warn("Couldn't get quiesce db leader from the mds map")
+                self.sleep_unless_stopped(5)
+
+        while leader is not None and not self.is_stopped:
+            # We use the one_shot here to cover for cases when the mds crashes
+            # without this parameter the client may get stuck awaiting response from a dead MDS
+            command = ['tell', f"mds.{leader}", 'quiesce', 'db']
+            command.extend(args)
+            self.logger.debug("Running ceph command: '%s'" % " ".join(command))
+            result = self.fs.run_ceph_cmd(args=command, check_status=False, stdout=StringIO())
+            rc, stdout = result.exitstatus, result.stdout.getvalue()
+            if rc == -errno.ENOTTY:
+                try:
+                    resp = json.loads(stdout)
+                    leader = int(resp['leader'])
+                    self.logger.info("Retrying a quiesce db command with leader %d" % leader)
+                except Exception as e:
+                    self.logger.error("Couldn't parse ENOTTY response from an mds with error: %s\n%s" % (str(e), stdout))
+                    self.sleep_unless_stopped(5)
+            else:
+                break
+
+        return (rc, stdout)
+    
+    def get_set_state_name(self, response, set_id = None):
+        if isinstance(response, (str, bytes, bytearray)):
+            response = json.loads(response)
+
+        sets = response['sets']
+        if len(sets) == 0:
+            raise ValueError("response has no sets")
+
+        if set_id is None:
+            if len(sets) > 1:
+                raise ValueError("set_id must be provided for a multiset response")
+            else:
+                set_id = next(iter(sets.keys()))
+
+        return response['sets'][set_id]['state']['name']
+
+    def check_canceled(self, response, set_id = None):
+        if 'CANCELED' == self.get_set_state_name(response, set_id):
+            self.logger.warn('''
+                             Quiesce set got cancelled. Won't raise an error since this could be a failover, 
+                             will wait for the next quiesce attempt''')
```

Comment:
If you carried `status` through to this method you could check for failover and fail otherwise.

Reply:
I didn't want to start digging into whether a failover or some other event had occurred. We just don't treat CANCELED as a fatal failure, log about it and move on. Other tasks running in parallel will decide whether a failover or whatever else was fatal.